### DISCUSSION TITLE
RP2040 - Pi Pico W - Micropython- Wifi & bluetooth connection issues.

### DISCUSSION THREAD

**Discussion by Iscandar77** [micropython/micropython#13615] [RP2040 / Pico]
Good evening everyone,

I am encountering difficulties in properly configuring my Raspberry Pi Pico W for both Bluetooth and Wi-Fi connections. Below, I detail the steps I've followed and the issues encountered, hoping someone might offer a solution or suggestions.

**Environment:**

Hardware: Raspberry Pi Pico W.
Firmware: version 1.22.1 (released on 01/05/2024).
IDE: Thonny version 4.1.4.
Operating System: Ubuntu 23.04,  but testedeven  on Windows 11 always with Thonny.
Guides followed: "Connecting to the Internet with Raspberry Pi Pico W" (updated 06/14/2023),(https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf)  Chapter 6 for Bluetooth in MicroPython.
Scripts tested: ble_advertising.py, picow_ble_temp_sensor.py.
Libraries: Attempted with and without the aioble package installed.

**Procedure and Issues Encountered:**
 I followed the instructions for Bluetooth connection step by step as indicated in the book.
 I downloaded and attempted to run the ble_advertising.py and picow_ble_temp_sensor.py scripts.
 I used two different Raspberry Pi Pico Ws and switched 2 different  USB cables to rule out hardware issues.

**The Bluetooth  connection does not work, receiving errors:**

```
>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 107, in <module>
  File "<stdin>", line 96, in demo
  File "<stdin>", line 43, in __init__
OSError: [Errno 22] EINVAL
```

 Testing the Pico W with a basic script to blink an LED, I encounter no problems, indicating the device functions correctly for simple operations.    Similar issues occur using both Ubuntu and Windows, suggesting the problem is not related to the operating system.

I tested WI-FI too:
- https://projects.raspberrypi.org/en/projects/get-started-pico-w/2 
- https://core-electronics.com.au/guides/raspberry-pi-pico-w-create-a-simple-http-server/
And got similar mistake, e.g. 

MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 19, in <module>
OSError: [Errno 1] EPERM

Questions:

    Could this be an issue related to permissions or library configuration?
    Has anyone experienced similar issues or has suggestions on how to solve these connection problems?
    Are there additional steps or checks I should perform to diagnose the issue?

Thank you in advance for any help or suggestion you can provide.
![Schermata del 2024-02-07 20-56-27](https://github.com/micropython/micropython/assets/104420134/7a560286-7323-41c9-9774-fcc07e4c5ebf)
![Schermata del 2024-02-07 20-55-54](https://github.com/micropython/micropython/assets/104420134/8c8c7d19-794f-45b9-8fd1-780df07acfc7)

Thank you for support.

**Comment by brianreinhold** [comment]
I can't say much about the wifi connections but I have done a lot with the Bluetooth and have no problems with connection. The nightmare is pairing as it has not been configured in the PICO-W port. Neither has Resolvable Privacy Addressing (RPA) which some of the examples in the MicroPython distro use. That error means there is something needed by your code that is not present. Probably advertising configured with adr_mode=2 (RPA) somewhere (just a guess). That will bork.

I don't know where you got your sample code. If I could look at it I could probably guess where it is going awry. I used the aioble examples too, but none of them pair. Wrote my own and connection works fine (central). I have lots of health devices I need a central for.

I scan for devices
filter the advertisements/scanresponses to find only devices I want to support (BT SIG health device profiles)
and then I connect to the device. 

Looks like this
```
async def connect(device: Device):
    print("Connecting to", device)
    try:
        connection: DeviceConnection | None = await device.connect()
        if (type(connection) != DeviceConnection):
            return
        print(type(connection._conn_handle))
```

Are you interested in a central or a peripheral? (Remember that it is the central that must initiate the connection.)

**Reply by Iscandar77** [reply]
Hello, essentially I'm a beginner who is trying to learn, so I'm operating for educational purposes and I would like to learn both central and peripheral aspects. The codes I've taken are the official ones from Raspberry (https://www.raspberrypi.com/), which is why I'm very disturbed that they are not working. Then, as mentioned, I've tried several from the internet. I will now try to modify the advertising and see. Thank you for the feedback. However, I fear it's more of a hardware problem at this point. I'll try buying 2 more Pico boards not from AliExpress, just to remove any doubt :)
